### PR TITLE
torch.tensor(..., device='meta') now creates an empty tensor on passing Ellipses with device='meta'

### LINKED ISSUE #55191: `torch.tensor(..., device='meta')` doesn't work
It can be made to work cuz ellipses is an actual Python object, just need to extend the Python binding code a little.

### PR DESCRIPTION
**Pull request by venkateshtata**
Fixes #55191

Screenshot 2021-05-12 at 8 02 53 AM

### CODE REVIEW

**Comment by ezyang** [comment]
Tests?

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -771,16 +773,53 @@ void _validate_sparse_coo_tensor_args(c10::DispatchKey dispatch_key, at::ScalarT
...
 
   constexpr int ctor_num_args = 6;
   ParsedArgs<ctor_num_args> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
+
   if (r.idx == 0) {
     PyObject* data = r.pyobject(0);
+
+    PyObject* meta_str = PyUnicode_DecodeFSDefault("meta");
+    PyObject* ellipsis_str = PyUnicode_DecodeFSDefault("Ellipsis");
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:794]: Just refer to Py_Ellipsis directly

File: torch/csrc/utils/tensor_new.cpp [hunk 2]
```diff
@@ -771,16 +773,53 @@ void _validate_sparse_coo_tensor_args(c10::DispatchKey dispatch_key, at::ScalarT
...
+
+    PyObject* data_passed = PyObject_Str(data);
+    PyObject* device_passed = r.pyobject(2);
+
+    if(PyObject_RichCompareBool(device_passed, meta_str, Py_EQ)==1 && PyObject_RichCompareBool(data_passed, ellipsis_str, Py_EQ)==1){
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:799]: This is a very bad way to do a device equality test. Instead parse the device into a at::Device and then do a normal equality. Ellipsis can be done via object identity test

File: torch/csrc/utils/tensor_new.cpp [hunk 3]
```diff
@@ -771,16 +773,53 @@ void _validate_sparse_coo_tensor_args(c10::DispatchKey dispatch_key, at::ScalarT
...
+        bool type_inference = r.isNone(1);
+        bool pin_memory = r.toBool(3);
+        auto sizes = compute_sizes(data);
+        Tensor new_tensor;
+        at::AutoDispatchBelowADInplaceOrView guard;  // TODO: remove
+        at::tracer::impl::NoTracerDispatchMode tracer_guard;
+        new_tensor = at::empty(0);
+        bool args_requires_grad = r.toBool(4);
+        new_tensor.detach_(); // ensure new_tensor a leaf node
+        new_tensor.set_requires_grad(args_requires_grad);
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:815]: You need to respect sizes. `tensor(..., device='meta', size=(2,))` should work too.

File: torch/csrc/utils/tensor_new.cpp [hunk 4]
```diff
@@ -208,6 +208,7 @@ Tensor internal_new_from_data(
     bool type_inference,
     bool pin_memory = false) {
 
+
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:211]: avoid adding random whitespace changes in PRs

**Review by ezyang — CHANGES_REQUESTED** [review]
tests, and the implementation is not quite right

File: torch/csrc/utils/tensor_new.cpp (hunk 4 above)
Comment (venkateshtata) [reply torch/csrc/utils/tensor_new.cpp:211]: Will keep that in mind. Thank you.

File: torch/csrc/utils/tensor_new.cpp (hunk 2 above)
Comment (venkateshtata) [reply torch/csrc/utils/tensor_new.cpp:799]: Screenshot 2021-05-13 at 7 38 27 AM

Can I perform checking like this ?

Comment (venkateshtata) [reply torch/csrc/utils/tensor_new.cpp:799]: Screenshot 2021-05-13 at 7 47 24 AM

Can I perform the check in this way ? I'm now using Py_Ellipsis to compare and considering size in the new tensor. Or is there any other method that I can use to perform object identity test ?

Comment (ezyang) [reply torch/csrc/utils/tensor_new.cpp]: Equals test is OK because I think ellipsis is a singleton/ String test is not OK.

Comment (venkateshtata) [reply torch/csrc/utils/tensor_new.cpp]: Screenshot 2021-05-13 at 3 23 21 PM

Will this be fine ? Using `at::Device` to compare as you suggested.

Could you tell me how to post my tests with the PR ? Thank you.

Comment (venkateshtata) [reply torch/csrc/utils/tensor_new.cpp]: Screenshot 2021-05-13 at 3 30 34 PM

Using `at::Device` for comparing as you suggested. Would that work ?

Could please tell me how to post my tests with the PR ? Thank you.

**Pull request closed without merging**